tools/src/main/java/org/apache/kafka/tools/consumer/ApiMessageFormatter.java

```diff
@@ -0,0 +1,99 @@
+ */
+package org.apache.kafka.tools.consumer;
+
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.common.MessageFormatter;
+import org.apache.kafka.common.protocol.ApiMessage;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.JsonNodeFactory;
+import com.fasterxml.jackson.databind.node.NullNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.fasterxml.jackson.databind.node.TextNode;
+
+import java.io.IOException;
+import java.io.PrintStream;
+import java.nio.ByteBuffer;
+import java.util.Objects;
+import java.util.Optional;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+public abstract class ApiMessageFormatter implements MessageFormatter {
+
+    private static final String VERSION = "version";
+    private static final String DATA = "data";
+    private static final String KEY = "key";
+    private static final String VALUE = "value";
+    static final String UNKNOWN = "unknown";
+
+    @Override
+    public void writeTo(ConsumerRecord<byte[], byte[]> consumerRecord, PrintStream output) {
+        ObjectNode json = new ObjectNode(JsonNodeFactory.instance);
+
+        byte[] key = consumerRecord.key();
+        if (Objects.nonNull(key)) {
+            short keyVersion = ByteBuffer.wrap(key).getShort();
+            JsonNode dataNode = readToKeyNode(ByteBuffer.wrap(key), keyVersion);
+
+            if (dataNode instanceof NullNode) {
+                return;
+            }
+            json.putObject(KEY)
+                    .put(VERSION, keyVersion)
+                    .set(DATA, dataNode);
+        } else {
+            json.set(KEY, NullNode.getInstance());
+        }
+
+        byte[] value = consumerRecord.value();
+        if (Objects.nonNull(value)) {
+            short valueVersion = ByteBuffer.wrap(value).getShort();
+            JsonNode dataNode = readToValueNode(ByteBuffer.wrap(value), valueVersion);
+
+            json.putObject(VALUE)
+                    .put(VERSION, valueVersion)
+                    .set(DATA, dataNode);
+        } else {
+            json.set(VALUE, NullNode.getInstance());
+        }
+
+        try {
+            output.write(json.toString().getBytes(UTF_8));
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private JsonNode readToKeyNode(ByteBuffer byteBuffer, short version) {
+        return readToKeyMessage(byteBuffer)
+                .map(logKey -> transferKeyMessageToJsonNode(logKey, version))
+                .orElseGet(() -> new TextNode(UNKNOWN));
+    }
+    
+    private JsonNode readToValueNode(ByteBuffer byteBuffer, short version) {
+        return readToValueMessage(byteBuffer)
+                .map(logValue -> transferValueMessageToJsonNode(logValue, version))
+                .orElseGet(() -> new TextNode(UNKNOWN));
+    }
+    
+    protected abstract Optional<ApiMessage> readToKeyMessage(ByteBuffer byteBuffer);
```

Comment:
It seems `readToKeyMessage`  and `transferKeyMessageToJsonNode` cab be merged into single method. For example: `protected abstract Optional<JsonNode> readToKeyJson(ByteBuffer byteBuffer, short version);`